Bundle C++ to SDK to artifact that can be used with CMake's fetch_content (#3875)
rerun-io/rerun: rerun_cpp/CMakeLists.txt

add_library(rerun_sdk ${rerun_sdk_SRC})

> # Rerun needs C++17.

Reviewer: ```suggestion
# Rerun needs at least C++17.
```

set_property(TARGET rerun_sdk PROPERTY CXX_STANDARD 17)

# ------------------------------------------------------------------------------
# Rerun development setup only:
if(DEFINED RERUN_REPOSITORY)
    # execute process doesn't fail if the process fails.
    # `COMMAND_ERROR_IS_FATAL ANY` parameter fixes this but is only available in CMake 3.19
    # which isn't default on Ubuntu LTS as of writing, causing unnecessary friction.
    # TODO(andreas): use add_custom_command instead so this runs at build time! https://cmake.org/cmake/help/latest/command/add_custom_command.html#command:add_custom_command

>     execute_process(COMMAND cargo build --release -p re_types RESULT_VARIABLE ret) # Generates most of the C++ source files

Reviewer: ```suggestion
    execute_process(COMMAND just codegen RESULT_VARIABLE ret) # Generates most of the C++ source files
```

Author: It makes sense, but I'm unhappy about this requiring more dependencies for people coming in to build cpp themselves

Author: actually I think now we should just remove this line. Codegen should *always* be explicit now

    if(NOT(ret EQUAL "0"))
        message(FATAL_ERROR "Failed to build re_types in order to generate C++ source files.")
    endif()

    execute_process(COMMAND cargo build --release -p rerun_c RESULT_VARIABLE ret) # We link against this, so must be up-to-date

    if(NOT(ret EQUAL "0"))
        message(FATAL_ERROR "Failed to build rerun_c.")
    endif()

    # Overwrite where to find rerun_c library.
    if(APPLE)
        set(RERUN_C_LIB_DEFAULT ${CMAKE_CURRENT_SOURCE_DIR}/../target/release/librerun_c.a)
    elseif(UNIX) # if(LINUX) # CMake 3.25
        set(RERUN_C_LIB_DEFAULT ${CMAKE_CURRENT_SOURCE_DIR}/../target/release/librerun_c.a)
    elseif(WIN32)
        set(RERUN_C_LIB_DEFAULT ${CMAKE_CURRENT_SOURCE_DIR}/../target/release/rerun_c.lib)
    else()
        message(FATAL_ERROR "Unsupported platform.")
    endif()

    # Set very strict warning settings when we're testing the SDK.
    # We don't want to force this on any user!
    set_default_warning_settings(rerun_sdk)

>     # Put Rerun.h into the same place where it's on a user's machine and apply CMake variables like version number.

Reviewer: ```suggestion
    # Put `rerun.h` into the same place where it's on a user's machine and apply CMake variables like version number.
```
